core/lib/web3_decl/src/client.rs

```diff
@@ -0,0 +1,456 @@
+//! L2 HTTP client.
+
+use std::{
+    fmt,
+    num::NonZeroUsize,
+    sync::{Arc, Mutex},
+    time::Duration,
+};
+
+use async_trait::async_trait;
+use jsonrpsee::{
+    core::{
+        client::{BatchResponse, ClientT, Error, Subscription, SubscriptionClientT},
+        params::BatchRequestBuilder,
+        traits::ToRpcParams,
+    },
+    http_client::{HttpClient, HttpClientBuilder},
+};
+use serde::de::DeserializeOwned;
+use tokio::time::Instant;
+
+#[derive(Debug, Clone, Copy)]
+enum RateLimitOrigin<'a> {
+    Notification(&'a str),
+    Request(&'a str),
+    BatchRequest(&'a BatchRequestBuilder<'a>),
+    Subscription(&'a str),
+}
+
+impl fmt::Display for RateLimitOrigin<'_> {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Notification(name) => write!(formatter, "notification `{name}`"),
+            Self::Request(name) => write!(formatter, "request `{name}`"),
+            Self::BatchRequest(req) => {
+                let method_names: Vec<_> = req.iter().map(|(name, _)| name).collect();
+                write!(formatter, "batch request {method_names:?}")
+            }
+            Self::Subscription(name) => write!(formatter, "subscription `{name}`"),
+        }
+    }
+}
+
+/// JSON-RPC client for the main node with built-in middleware support.
+///
+/// The client should be used instead of `HttpClient` etc. A single instance of the client should be built
+/// and shared among all tasks run by the node in order to correctly rate-limit requests.
+#[derive(Debug, Clone)]
+pub struct L2Client {
```

Comment:
L2Client -> RateLimittedHttpClient as it's a wrapper for HttpClient?

Reply:
I've went with a more general name because we might want to introduce new middleware in the future and/or support multiple transports for the client (e.g., WS).